src/io/frame.rs

```diff
@@ -243,7 +243,7 @@ pub trait Codec {
     /// This method will encode `msg` into the byte buffer provided by `buf`.
     /// The `buf` provided is an internal buffer of the `Framed` instance and
     /// will be written out when possible.
-    fn encode(&mut self, msg: Self::Out, buf: &mut Vec<u8>);
+    fn encode(&mut self, msg: Self::Out) -> Result<Vec<u8>, io::Error>;
```

Comment:
Here it's actually important to take a destination buffer as an argument to avoid double-buffering or copying data unnecessarily. Perhaps this could still take `&mut Vec<u8>` but it could return `io::Result<()>`?